vyper/builtin_functions/functions.py

```diff
@@ -2156,8 +2158,98 @@ def build_IR(self, expr, context):
         )
 
 
+class ABIDecode(_SimpleBuiltinFunction):
+    _id = "_abi_decode"
+    _inputs = [("data", BytesArrayPrimitive()), ("output_type", "TYPE_DEFINITION")]
+    _kwargs = {"unwrap_tuple": KwargSettings(BoolDefinition(), True, require_literal=True)}
+
+    def fetch_call_return(self, node):
+        _, output_type = self.infer_arg_types(node)
+        return output_type.typedef
+
+    def infer_arg_types(self, node):
+        validate_call_args(node, 2, ["unwrap_tuple"])
+
+        data_type = get_exact_type_from_node(node.args[0])
+        output_typedef = TypeTypeDefinition(
+            get_type_from_annotation(node.args[1], DataLocation.MEMORY)
+        )
+
+        return [data_type, output_typedef]
+
+    @validate_inputs
+    def build_IR(self, expr, args, kwargs, context):
+        unwrap_tuple = kwargs["unwrap_tuple"]
+
+        data = args[0]
+        output_typ = args[1]
+        wrapped_typ = output_typ
+
+        if unwrap_tuple is True:
+            wrapped_typ = calculate_type_for_external_return(output_typ)
+
+        abi_size_bound = wrapped_typ.abi_type.size_bound()
+        abi_min_size = wrapped_typ.abi_type.min_size()
+
+        # Get the size of data
+        input_max_len = data.typ.maxlen
+
```

Comment:
maybe add a sanity check -
```suggestion

        assert abi_min_size <= abi_size_bound, "bad abi type"
```